pkg/v1beta1/cache/cache.go

```diff
@@ -0,0 +1,166 @@
+package cache
+
+import (
+	"fmt"
+	"strings"
+
+	"sync"
+
+	"github.com/go-logr/logr"
+	v1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+	klog "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// GatewayLookupCache provides concurrency safe lookups from dns host to namespace/gateway
+// Use New() as the Add, Delete, and Get functions all assume the cache map is non-nil
+type GatewayLookupCache struct {
+	cache  map[string]string
+	mutex  sync.Mutex
+	logger logr.Logger
+}
+
+func New() *GatewayLookupCache {
+	return &GatewayLookupCache{
+		cache:  make(map[string]string),
+		mutex:  sync.Mutex{},
+		logger: klog.Log,
+	}
+
+}
+
+func (glc *GatewayLookupCache) Add(gateway string, hosts ...string) {
+	glc.mutex.Lock()
+	defer glc.mutex.Unlock()
+
+	for _, host := range hosts {
+		glc.cache[host] = gateway
+	}
+}
+
+func (glc *GatewayLookupCache) Delete(hosts ...string) {
+	glc.mutex.Lock()
+	defer glc.mutex.Unlock()
+
+	for _, host := range hosts {
+		delete(glc.cache, host)
+	}
+}
+
+func (glc *GatewayLookupCache) Get(host string) (string, bool) {
+	glc.mutex.Lock()
+	defer glc.mutex.Unlock()
+
+	gw, ok := glc.cache[host]
+	return gw, ok
```

Comment:
```suggestion
	return glc.cache[host]
```

Reply:
That would only return a string vs a string and bool. I suppose you could test against the empty string but Get is designed to be clear if there is a cache miss vs a default result.

Reply:
ok. You are right. Because its a map and its strange.  nvm